src/Servers/Kestrel/Core/src/Internal/Http3/Http3HeadersEnumerator.cs

```diff
@@ -0,0 +1,152 @@
+                    break;
+                default:
+                    _headersType = HeadersType.Untyped;
+                    _genericEnumerator = headers.GetEnumerator();
+                    break;
+            }
+
+            _hasMultipleValues = false;
+        }
+
+        public bool MoveNext()
+        {
+            if (_hasMultipleValues && MoveNextOnStringEnumerator(Current.Key))
+            {
+                return true;
+            }
+
+            if (_headersType == HeadersType.Headers)
+            {
+                return _headersEnumerator.MoveNext()
+                    ? SetCurrent(_headersEnumerator.Current.Key, _headersEnumerator.Current.Value, _headersEnumerator.CurrentKnownType)
+                    : false;
+            }
+            else if (_headersType == HeadersType.Trailers)
+            {
+                return _trailersEnumerator.MoveNext()
+                    ? SetCurrent(_trailersEnumerator.Current.Key, _trailersEnumerator.Current.Value, _trailersEnumerator.CurrentKnownType)
+                    : false;
+            }
+            else
+            {
+                return _genericEnumerator!.MoveNext()
+                    ? SetCurrent(_genericEnumerator.Current.Key, _genericEnumerator.Current.Value, default)
+                    : false;
+            }
+        }
+
+        private bool MoveNextOnStringEnumerator(string key)
+        {
+            var result = _stringValuesEnumerator.MoveNext();
+            Current = result ? new KeyValuePair<string, string>(key, _stringValuesEnumerator.Current) : default;
+            return result;
+        }
+
+        private bool SetCurrent(string name, StringValues value, KnownHeaderType knownHeaderType)
+        {
+            _knownHeaderType = knownHeaderType;
+
+            if (value.Count == 1)
+            {
+                Current = new KeyValuePair<string, string>(name, value.ToString());
+                _hasMultipleValues = false;
+                return true;
+            }
+            else
+            {
+                _stringValuesEnumerator = value.GetEnumerator();
+                _hasMultipleValues = true;
+                return MoveNextOnStringEnumerator(name);
+            }
+        }
+
+        public void Reset()
+        {
+            if (_headersType == HeadersType.Headers)
+            {
+                _headersEnumerator.Reset();
+            }
+            else if (_headersType == HeadersType.Trailers)
+            {
+                _trailersEnumerator.Reset();
+            }
+            else
+            {
+                _genericEnumerator!.Reset();
+            }
+            _stringValuesEnumerator = default;
+            _knownHeaderType = default;
+        }
+
+        public void Dispose()
+        {
+        }
+
+        internal static int GetResponseHeaderStaticTableId(KnownHeaderType responseHeaderType)
+        {
+            // Not Implemented
```

Comment:
When will we need to do this? Do we need to file a followup issue?

Reply:
https://github.com/dotnet/aspnetcore/issues/33980